src/Interval.cpp

```diff
@@ -157,5 +157,108 @@ Expr Interval::neg_inf_noinline() {
     return Interval::neg_inf_expr;
 }
 
+ConstantInterval::ConstantInterval() = default;
+
+ConstantInterval::ConstantInterval(int64_t min, int64_t max)
+    : min(min), max(max), min_defined(true), max_defined(true) {
+    internal_assert(min <= max);
+}
+
+ConstantInterval ConstantInterval::everything() {
+    return ConstantInterval();
+}
+
+ConstantInterval ConstantInterval::single_point(int64_t x) {
+    return ConstantInterval(x, x);
+}
+
+ConstantInterval ConstantInterval::bounded_below(int64_t min) {
+    ConstantInterval result(min, min);
+    result.max_defined = false;
+    return result;
+}
+
+ConstantInterval ConstantInterval::bounded_above(int64_t max) {
+    ConstantInterval result(max, max);
+    result.min_defined = false;
+    return result;
+}
+
+bool ConstantInterval::is_everything() const {
+    return !min_defined && !max_defined;
+}
+
+bool ConstantInterval::is_single_point() const {
+    return !is_everything() && min == max;
```

Comment:
I don't think `!is_everything()` is sufficient, because any partially bounded interval with the same `min` and `max` will be defined as a single point.

i.e.:
```
ConstantInterval c = ConstantInterval::bounded_below(4);
c.is_single_point(); // -> true
```
Pretty sure it should be:
```
min_defined && max_defined && min == max;
```